clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/customtemplates/CustomTemplate.kt

```diff
@@ -0,0 +1,271 @@
+package com.clevertap.android.sdk.inapp.customtemplates
+
+import com.clevertap.android.sdk.inapp.customtemplates.CustomTemplateType.FUNCTION
+import com.clevertap.android.sdk.inapp.customtemplates.CustomTemplateType.TEMPLATE
+import com.clevertap.android.sdk.inapp.customtemplates.TemplateArgumentType.ACTION
+import com.clevertap.android.sdk.inapp.customtemplates.TemplateArgumentType.BOOLEAN
+import com.clevertap.android.sdk.inapp.customtemplates.TemplateArgumentType.FILE
+import com.clevertap.android.sdk.inapp.customtemplates.TemplateArgumentType.NUMBER
+import com.clevertap.android.sdk.inapp.customtemplates.TemplateArgumentType.STRING
+
+/**
+ * A definition of a custom template. Can be a function or a code template.
+ * Instances are uniquely identified by their name.
+ */
+class CustomTemplate private constructor(
+    /**
+     * The name of the template. Must be unique and non-blank.
+     */
+    val name: String,
+
+    /**
+     * Whether the template has UI or not. If set to `true` the template is registered as part of the in-apps queue
+     * and must be explicitly dismissed before other in-apps can be shown. If set to `false` the template is executed
+     * directly and does not require dismissal nor it impedes other in-apps.
+     */
+    val isVisual: Boolean,
+
+    internal val args: List<TemplateArgument>,
+    internal val type: CustomTemplateType
+) {
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (javaClass != other?.javaClass) return false
+
+        other as CustomTemplate
+
+        return name == other.name
+    }
+
+    override fun hashCode(): Int {
+        return name.hashCode()
+    }
+
+    /**
+     * Builder for [CustomTemplate] functions. Function arguments' names cannot contain "."
+     * @param isVisual See [CustomTemplate.isVisual]
+     */
+    class FunctionBuilder(isVisual: Boolean) :
+        Builder<FunctionBuilder>(FUNCTION, allowHierarchicalNames = false, isVisual) {
+
+        override val thisRef: FunctionBuilder = this
```

Comment:
why not initialise in const itself?

Reply:
`this` ref cannot be passed in the super constructor invocation, because it is not initialized at that point.